sweep/sweeper_test.go

```diff
@@ -0,0 +1,954 @@
+	remoteTx := &wire.MsgTx{
+		TxIn: []*wire.TxIn{
+			{
+				PreviousOutPoint: *(spendableInputs[0].OutPoint()),
+			},
+		},
+	}
+	err = ctx.backend.publishTransaction(remoteTx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if postSweep {
+		ctx.tick()
+
+		// Tx publication by sweeper returns ErrDoubleSpend. Sweeper
+		// will retry the inputs without reporting a result. It could be
+		// spent by the remote party.
+		ctx.receiveTx()
+	}
+
+	ctx.backend.mine()
+
+	select {
+	case result := <-resultChan1:
+		if result.Err != ErrRemoteSpend {
+			t.Fatalf("expected remote spend")
+		}
+		if result.Tx.TxHash() != remoteTx.TxHash() {
+			t.Fatalf("expected remote spend tx")
+		}
+	case <-time.After(5 * time.Second):
+		t.Fatalf("no result received")
+	}
+
+	if !postSweep {
+		// Assert that the sweeper sweeps the remaining input.
+		ctx.tick()
+		sweepTx := ctx.receiveTx()
+
+		if len(sweepTx.TxIn) != 1 {
+			t.Fatal("expected sweep to only sweep the one remaining output")
+		}
+
+		ctx.backend.mine()
+
+		select {
+		case result := <-resultChan2:
+			if result.Err != nil {
+				t.Fatalf("expected sweep success")
+			}
+			if result.Tx.TxHash() != sweepTx.TxHash() {
+				t.Fatalf("expected sweep tx")
+			}
+		case <-time.After(5 * time.Second):
+			t.Fatalf("no result received")
+		}
+
+		ctx.finish(1)
+	} else {
+		// Expected sweeper to be still listening for spend of the
+		// error input.
+		ctx.finish(2)
+
+		select {
+		case <-resultChan2:
+			t.Fatalf("no result expected for error input")
+		default:
+		}
+	}
+}
+
+// TestIdempotency asserts that offering the same input multiple times is
+// handled correctly.
+func TestIdempotency(t *testing.T) {
+	ctx := createSweeperTestContext(t)
+
+	resultChan1, err := ctx.sweeper.SweepInput(spendableInputs[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	resultChan2, err := ctx.sweeper.SweepInput(spendableInputs[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx.tick()
+
+	ctx.receiveTx()
+
+	resultChan3, err := ctx.sweeper.SweepInput(spendableInputs[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Spend the input of the sweep tx.
+	ctx.backend.mine()
+
+	resultsReceived := 0
+	for resultsReceived < 3 {
+		select {
+		case result := <-resultChan1:
+			if result.Err != nil {
+				t.Fatalf("expected successful sweep")
+			}
+			resultsReceived++
+		case result := <-resultChan2:
+			if result.Err != nil {
+				t.Fatalf("expected successful sweep")
+			}
+			resultsReceived++
+		case result := <-resultChan3:
+			if result.Err != nil {
+				t.Fatalf("expected successful sweep")
+			}
+			resultsReceived++
+		case <-time.After(defaultTestTimeout):
+			t.Fatalf("no result received")
+		}
+	}
+
+	resultChan4, err := ctx.sweeper.SweepInput(spendableInputs[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Because the sweeper isn't keeping a history of all spend inputs, it
+	// cannot recognize that it already swept this input itself!
+	select {
+	case result := <-resultChan4:
+		if result.Err != nil {
```

Comment:
Shouldn't this actually fail with a double spend error? Assuming either the test backend keeps a proper mempool, or a block is mined before we try to sweep this input for the 4th time.

Reply:
No, not with a double spend. It should be idempotent, so we expect a success. We swept it already. 

Updated comments to reflect what is actually being tested, was outdated.